### PR TITLE
Feat/use token balance

### PR DESCRIPTION
**Pull request by AlexNi245**
## Description
A simple hook that fetches the user's balance of a certain ERC20 token. 

## Aditional Information
I created 3 testcases in the storybook section
1. Success -> Owner and Contract Id's are valid
2. Error -> undefined contract address
3. Error -> undefined owner address

I want to write unit test in the first place but still not sure how to mock the web3context within a unit test :( so I decided to used the storybook instead

### CODE REVIEW

File: packages/hooks/src/hooks/useTokenBalance/ERC20.json
Comment (etr2460) [inline packages/hooks/src/hooks/useTokenBalance/ERC20.json:1]: Did you write this, or was it generated/sourced from elsewhere? I wonder if we can import this from a library instead of needing to define the schema for an ERC20 contract

File: packages/hooks/src/hooks/useTokenBalance/index.ts
Comment (etr2460) [inline packages/hooks/src/hooks/useTokenBalance/index.ts:1]: i think you also need to export this from the root hooks directory here: https://github.com/Developer-DAO/web3-ui/blob/master/packages/hooks/src/index.ts

File: packages/hooks/src/hooks/useTokenBalance/useTokenBalance.ts
Comment (etr2460) [inline packages/hooks/src/hooks/useTokenBalance/useTokenBalance.ts:6]: not sure what naming conventions are common here, but maybe `contractAddress`? I could also see us wanting `ownerAddress` instead of `owner`, unless we also accept ENS domains for the owner string

File: packages/hooks/src/hooks/useTokenBalance/useTokenBalance.ts
Comment (etr2460) [inline packages/hooks/src/hooks/useTokenBalance/useTokenBalance.ts:8]: This should be fine, and will default `balance` to `undefined`
```suggestion
  const [balance, setBalance] = React.useState<string>();
```

File: packages/hooks/src/hooks/useTokenBalance/useTokenBalance.ts
Comment (etr2460) [inline packages/hooks/src/hooks/useTokenBalance/useTokenBalance.ts:9]: maybe it's better to make this `errorMessage` and set the type to `string`. Then, depending on the error, we can pass a different message up

File: packages/hooks/src/stories/UseTokenBalance.stories.tsx
Comment (etr2460) [inline packages/hooks/src/stories/UseTokenBalance.stories.tsx:30]: can we pull these strings into constants at the top of the file (since they're reused) and name them appropriately? I assume that `address` is `DAI_CONTRACT_ADDRESS`?

File: packages/hooks/src/stories/UseTokenBalance.stories.tsx
Comment (etr2460) [inline packages/hooks/src/stories/UseTokenBalance.stories.tsx:23]: should we use a non mainnet for stories? I think we used rinkeby for other stories

**Review by etr2460 — COMMENTED** [review]
a few comments. i think it's probably ok to not have unit tests for now since you added the storybook. but i'll need to look into other possible tests at some point. there's got to be a good way to mock the provider

File: packages/hooks/src/hooks/useTokenBalance/ERC20.json
Comment (AlexNi245) [reply packages/hooks/src/hooks/useTokenBalance/ERC20.json:1]: i copied it from the following  [sitw](https://ethereumdev.io/abi-for-erc20-contract-on-ethereum/) but I've you know a library that provides the interface we should replace the .json file

File: packages/hooks/src/hooks/useTokenBalance/useTokenBalance.ts
Comment (AlexNi245) [reply packages/hooks/src/hooks/useTokenBalance/useTokenBalance.ts:6]: I think the obvious the function names the better. So I also think that  ownerAddress and contractAdrees are the better options for this. 
But haven't tested the hook with an ens domain yet so I don't know how it behaves in this case.

Comment (etr2460) [reply packages/hooks/src/hooks/useTokenBalance/useTokenBalance.ts:6]: it probably doesn't support ENS unless you add it. I think we've implemented ENS support in other hooks/components already, so maybe it makes sense to pull that code out into a util/hook and reuse it here too. ENS support isn't blocking though

File: packages/hooks/src/hooks/useTokenBalance/ERC20.json
Comment (etr2460) [reply packages/hooks/src/hooks/useTokenBalance/ERC20.json:1]: I looked around, but couldn't find a legit looking library for this anywhere (i found one with only 1 star, but that seems risky to rely on).

Maybe let's rename the file to `ERC20ABI.json` at least, for a bit more clarity? I assume it'll never need to change (i'm still a bit new to web3, so don't have a full understanding of this yet)

Comment (AlexNi245) [reply packages/hooks/src/hooks/useTokenBalance/ERC20.json:1]: Yes the ERC20 interface will not change over time. Will rename the file to achieve more clearity

File: packages/hooks/src/hooks/useTokenBalance/useTokenBalance.ts
Comment (AlexNi245) [reply packages/hooks/src/hooks/useTokenBalance/useTokenBalance.ts]: i think util for ens would make sense here. I can have a look at how to do this best after this one is merged.

File: packages/hooks/src/hooks/useTokenBalance/useTokenBalance.ts
Comment (Dhaiwat10) [inline packages/hooks/src/hooks/useTokenBalance/useTokenBalance.ts:14]:
```suggestion
  const [error, setError] = React.useState(false);
```

I think `error` is a better name

File: packages/hooks/src/hooks/useTokenBalance/useTokenBalance.ts
Comment (Dhaiwat10) [inline packages/hooks/src/hooks/useTokenBalance/useTokenBalance.ts:12]:
```suggestion
  const [loading, setLoading] = React.useState(true);
```

File: packages/hooks/src/hooks/useTokenBalance/useTokenBalance.ts
Comment (Dhaiwat10) [inline packages/hooks/src/hooks/useTokenBalance/useTokenBalance.ts:51]:
```suggestion
  return { balance, loading, error };
```

Let's follow the convention used in the other hook!

**Review by Dhaiwat10 — CHANGES_REQUESTED** [review]
Thanks for this PR @AlexNi245! I have requested a few changes. Can you please look into them?

File: packages/hooks/src/hooks/useTokenBalance/useTokenBalance.ts
Comment (AlexNi245) [reply packages/hooks/src/hooks/useTokenBalance/useTokenBalance.ts:14]: Is it okay for you @etr2460 ? If so I will rename it to "error"

**Comment by Dhaiwat10** [comment]
@AlexNi245 the storybook setup needs some changes. I'm gonna be working on it and pushing commits to this PR. So if you're working on this PR rn, please keep that on hold!

File: packages/hooks/src/hooks/useTokenBalance/useTokenBalance.ts
Comment (Dhaiwat10) [inline packages/hooks/src/hooks/useTokenBalance/useTokenBalance.ts:41]: @AlexNi245 this is one thing that I want you to look at. I just fixed this but you didn't have a dependency array attached to your `useEffect` which was causing it to run in an infinite loop.

File: packages/hooks/src/hooks/useTokenBalance/useTokenBalance.ts
Comment (Dhaiwat10) [inline packages/hooks/src/hooks/useTokenBalance/useTokenBalance.ts:49]: I added these two additional helper variables.

File: tsconfig.json
Comment (Dhaiwat10) [inline tsconfig.json:21]: Added this to support `.json` imports

File: packages/hooks/src/hooks/useTokenBalance/useTokenBalance.ts
Comment (etr2460) [reply packages/hooks/src/hooks/useTokenBalance/useTokenBalance.ts]: My thought was that this shouldn't be a boolean at all, but instead be an errorMessage and a string type, so that we can pass a better error message back to the caller:
>maybe it's better to make this errorMessage and set the type to string. Then, depending on the error, we can pass a different message up


I can iterate on this in a future PR though

**Comment by Dhaiwat10** [comment]
@allcontributors please add @AlexNi245 for code

**Pull request merged**
